web3/contract.py

```diff
@@ -926,157 +1385,62 @@ def call(
-            call_transaction: TxParams = {}
-        else:
-            call_transaction = cast(TxParams, dict(**transaction))
-
-        if 'data' in call_transaction:
-            raise ValueError("Cannot set 'data' field in call transaction")
-
-        if self.address:
-            call_transaction.setdefault('to', self.address)
-        if self.w3.eth.default_account is not empty:
-            # type ignored b/c check prevents an empty default_account
-            call_transaction.setdefault('from', self.w3.eth.default_account)  # type: ignore
-
-        if 'to' not in call_transaction:
-            if isinstance(self, type):
-                raise ValueError(
-                    "When using `Contract.[methodtype].[method].call()` from"
-                    " a contract factory you "
-                    "must provide a `to` address with the transaction"
-                )
-            else:
-                raise ValueError(
-                    "Please ensure that this contract instance has an address."
-                )
+        call_transaction = self._get_call_txparams(transaction)
 
-        block_id = parse_block_identifier(self.w3, block_identifier)
+        block_id = await async_parse_block_identifier(self.w3, block_identifier)
 
-        return call_contract_function(
+        return await async_call_contract_function(
             self.w3,
             self.address,
             self._return_data_normalizers,
             self.function_identifier,
             call_transaction,
-            block_id,
+            block_id,  # type: ignore
             self.contract_abi,
             self.abi,
             state_override,
             *self.args,
             **self.kwargs
         )
 
-    def transact(self, transaction: Optional[TxParams] = None) -> HexBytes:
-        if transaction is None:
-            transact_transaction: TxParams = {}
-        else:
-            transact_transaction = cast(TxParams, dict(**transaction))
-
-        if 'data' in transact_transaction:
-            raise ValueError("Cannot set 'data' field in transact transaction")
-
-        if self.address is not None:
-            transact_transaction.setdefault('to', self.address)
-        if self.w3.eth.default_account is not empty:
-            # type ignored b/c check prevents an empty default_account
-            transact_transaction.setdefault('from', self.w3.eth.default_account)  # type: ignore
-
-        if 'to' not in transact_transaction:
-            if isinstance(self, type):
-                raise ValueError(
-                    "When using `Contract.transact` from a contract factory you "
-                    "must provide a `to` address with the transaction"
-                )
-            else:
-                raise ValueError(
-                    "Please ensure that this contract instance has an address."
-                )
+    @classmethod
+    def factory(cls, class_name: str, **kwargs: Any) -> 'AsyncContractFunction':
```

Comment:
Same question here about combining the two `factory` definitions. I see the return type is different, but can we do a return type of `Union[ContractFunciton, AsyncContractFunction]` and combine the two `factory` definitions up into the Base class?

Reply:
done in #2466